### ISSUE TITLE
Wrong word-level timestamps in force-alignment starting from version 3.3.3

### ISSUE THREAD

**Issue by PavelPanjaya** [m-bain/whisperX#1220]
The word-level timestamps starting from version 3.3.3 are not accurate, i.e. not tightly aligned to the audio.
Adding an Audacity screenshot with timestamps versions made with 3.3.0 - 3.4.2 WhisperX force alignment model.

<img width="1498" height="720" alt="Image" src="https://github.com/user-attachments/assets/2dc1d7a6-0b28-4761-8218-c97bb7238142" />

**Comment by Dat-Ngu** [comment]
That's actually interesting. What are the implications of this issue? Less accurate transcription? More WER?

**Comment by PavelPanjaya** [comment]
I am using WhisperX for force-alignment only, as I already have the audio and the transcript. The implication of this are incorrect timestamps of the words.

**Comment by avoss** [comment]
I have the same problem using WhisperX 3.7.4. It's not just word-level timestamps but also segment-level. I found a simple reproduceable example, where Jeff Sachs speaks at the UN. He is introduced in Chinese and then starts to speak in English after 8 seconds. In the generated audio.json, the segment with his words starts immediately. When I replace the Chinese introduction with silence, then the segment starts correctly at 8 seconds.

So to me it looks like WhisperX uses the timestamp when it hears something that sounds like speech, but does not use the timestamp when it starts to understand something.

Maybe there are some cli arguments to control how noise is considered speech or something similar?

[audio-original.json](https://github.com/user-attachments/files/23151305/audio-original.json)
[audio-original.mp3](https://github.com/user-attachments/files/23151306/audio-original.mp3)
[audio-silence.json](https://github.com/user-attachments/files/23151307/audio-silence.json)
[audio-silence.mp3](https://github.com/user-attachments/files/23151308/audio-silence.mp3)

**Comment by wrichert** [comment]
Root cause seems to be #986 - "support timestamp for numbers"

Commit `69281f3` is the first commit that breaks alignment. It was merged into v3.3.3.

### **Bisect results**

| Commit | Description | Result |
| :---- | :---- | :---- |
| `v3.3.2` | Baseline | GOOD (0.2s) |
| `69281f3` | support timestamps for numbers | BAD (7.6s) |
| `65b2332` | make align a bit faster | BAD |
| `4ebfb07` | make no beam consistent with backtrack | BAD |
| `f286e7f` | refactor: type hints/imports | GOOD (reverted on a parallel branch) |
| `024bc84` | consolidate segment data handling | GOOD |
| `2f93e02` | add SegmentData type | GOOD |
| `289eadf` | fix a merge error | BAD (re-merged \#986 changes) |
| `e7712f4` | update import paths | BAD |

Commits `f286e7f`-`2f93e02` were on a parallel refactoring branch that hadn't yet merged the \#986 changes. The merge commit `289eadf` brought the \#986 code back in.

**What #986 changed in `alignment.py`**

Three interrelated changes that together break alignment:

1. Wildcard placeholders — Characters not in the wav2vec2 dictionary (numbers, punctuation) now get `*` placeholders instead of being skipped. These become token `-1`.  
1. Rewritten `get_trellis()` \+ new `get_wildcard_emission()` — For wildcard tokens, it takes `torch.max()` across ALL non-blank emission scores. This means wildcards eagerly match *any* speech signal, including Chinese/foreign language audio.  
1. `backtrack()` replaced with `backtrack_beam()` — Beam search finds a different (worse) path through the trellis for these cases.

The core problem: when there's non-target-language speech in the segment window, the wildcard emission scoring greedily matches it, pulling word timestamps back to the start of the foreign speech.  
@bfs18 - any chance you could have a second look at it? cc @m-bain

**Comment by PavelPanjaya** [comment]
Just a quick friendly reminder that the timestamps issue is still there. I’ve tested it again with WhisperX versions up to the latest one (3.8.1).

<img width="1899" height="738" alt="Image" src="https://github.com/user-attachments/assets/0bbd4223-b7e5-4f1f-90aa-e2cf3ddf3fd5" />

**Comment by m-bain** [comment]
@1carlito  can you look into this

**Comment by 1carlito** [comment]
Interesting, @Barabazs what have you tried so I can understand in more depth.

**Comment by wrichert** [comment]
@PavelPanjaya AFAIU @Barabazs only reverted it on his own fork. Can you try there whether it improved your case?

**Comment by PavelPanjaya** [comment]
@wrichert good point.

Here what I did:
```
git clone https://github.com/Barabazs/whisperX.git
cd whisperX
uv sync --all-extras --dev
```

And got whisperx version - afaiu I'm working with @Barabazs fork now:
```
+ whisperx==3.8.1 (from file:///<my_local_path>/whisperX)
```

timestamps still off:
<img width="1704" height="262" alt="Image" src="https://github.com/user-attachments/assets/15757792-0962-4121-831b-05f7e7df9ec3" />

**Comment by Barabazs** [comment]
@PavelPanjaya please share your audio sample and the exact command or script that you used. (need to replicate it)

@1carlito ignore my previous attempt. I was conflating this issue with #1315 

@wrichert I reverted my attempts because it didn't work. There's no point in retrying that.

**Comment by PavelPanjaya** [comment]
@Barabazs sure

[test_audio.mp3](https://github.com/user-attachments/files/25539538/test_audio.mp3)

(I see the same behavior in every audio with long pauses in the start and the end of the audio)

This is my code:

```
import whisperx
import torch
import numpy as np
from pathlib import Path
from importlib.metadata import version
ver = version("whisperx")

def to_float(x):
    if isinstance(x, np.generic):
        return float(x)
    return float(x)

def whisperx_to_audacity_labels(aligned_result, output_path):
    word_segments = aligned_result["word_segments"]

    with open(output_path, "w", encoding="utf-8") as f:
        for i, w in enumerate(word_segments):
            start = to_float(w["start"])
            end = to_float(w["end"])
            text = w["word"]
            f.write(f"{start}\t{end}\t{text}\n")

device = "cuda" if torch.cuda.is_available() else "cpu"

audio_path = Path("test_audio.mp3")
transcript = [
    {
        "text": "Hello world. This is a test.",
        "start": 0.0,
        "end": 8.0,
    }
]

audio = whisperx.load_audio(audio_path)

model_a, metadata = whisperx.load_align_model(
    language_code="en",
    device=device
)

aligned_result = whisperx.align(
    transcript,
    model_a,
    metadata,
    audio,
    device,
    return_char_alignments=False
)

labels_path = audio_path.parent / f"labels_{ver}.txt"
whisperx_to_audacity_labels(aligned_result, labels_path)
print(f"saved to: {labels_path}")
```

**Comment by avoss** [comment]
I posted a reproducible example a while back:

https://github.com/m-bain/whisperX/issues/1220#issuecomment-3448691057

Maybe that helps.

Thanks, Andreas

Am 25.02.2026 um 09:01 schrieb PavelPanjaya:

**Comment by Barabazs** [comment]
@PavelPanjaya why are you using this hardcoded transcript with incorrect start/end times?


> ```
> transcript = [
>     {
>         "text": "Hello world. This is a test.",
>         "start": 0.0,
>         "end": 8.0,
>     }
> ]
> ```

I don't have incorrect word timestamps when using your `whisperx_to_audacity_labels` on the latest version with the pipeline running from start to end.

**Comment by PavelPanjaya** [comment]
@Barabazs 
> why are you using this hardcoded transcript with incorrect start/end times?

It's my typical input for the force-aligner - a padded rough approximation of sentence start and end. In some cases the padding can be 1-2 seconds. Also in this specific example it's the start and the end of the audio.

> with the pipeline running from start to end.

By pipeline you mean Transcription + Force alignment?

**Comment by Barabazs** [comment]
> By pipeline you mean Transcription + Force alignment?

Yes.

Can you confirm whether you used this exact workflow and audio on v3.3.2 and it produced a correct alignment?


FWIW, with the example from @avoss, I do see the incorrect alignment in the beginning of the audio.

> So to me it looks like WhisperX uses the timestamp when it hears something that sounds like speech, but does not use the timestamp when it starts to understand something.

This seems more likely than what @wrichert wrote. 

@wrichert can you share your sample and workflow you used to bisect? The commit you found does not seem related to this bug/regression.

**Comment by wrichert** [comment]
I see. I used the files from https://github.com/m-bain/whisperX/issues/1220#issuecomment-3448691057 and the script below:

<details>
<summary>Bash bisection script</summary>

```bash

#!/bin/bash
# Bisect alignment.py changes between v3.3.2 and v3.3.3

WHISPERX_DIR="/Users/wrichert/src/prosody/test_1220/.venv313/lib/python3.13/site-packages/whisperx"
GIT_DIR="/Users/wrichert/src/prosody/test_1220/whisperx-repo.git"
TEST_SCRIPT="/Users/wrichert/src/prosody/test_1220/test_alignment.py"
VENV="/Users/wrichert/src/prosody/test_1220/.venv313"

test_commit() {
    local ref=$1
    local label=$2
    
    # Extract alignment.py at this commit
    git --git-dir="$GIT_DIR" show "$ref:whisperx/alignment.py" > "$WHISPERX_DIR/alignment.py" 2>/dev/null
    if [ $? -ne 0 ]; then
        echo "[$label] SKIP - no alignment.py at this commit"
        return 2
    fi
    
    # Fix imports: old versions use relative imports (.types, .audio, .utils) 
    # or may use whisperx.types instead of whisperx.schema
    # We already have a types.py shim, but also need to handle relative imports
    # that might conflict with the installed version's import structure
    
    # Clear cache
    rm -rf "$WHISPERX_DIR/__pycache__"
    
    # Run test
    output=$("$VENV/bin/python" "$TEST_SCRIPT" 2>&1)
    exit_code=$?
    
    # Extract relevant lines
    result=$(echo "$output" | grep -E "RESULT|First word")
    if [ -z "$result" ]; then
        # Show error
        error=$(echo "$output" | tail -3)
        echo "[$label ($ref)] ERROR: $error"
        return 2
    fi
    
    echo "[$label ($ref)] $result"
    return $exit_code
}

echo "=== Bisecting alignment.py commits between v3.3.2 and v3.3.3 ==="
echo ""

# The commits that changed alignment.py (in chronological order):
# 69281f3 support timestamps for numbers.
# 65b2332 make align a bit faster.
# 4ebfb07 make no beam consistent with backtrack.
# f286e7f refactor: improve type hints and clean up imports
# 024bc84 refactor: consolidate segment data handling in alignment function
# 2f93e02 feat: add SegmentData type for temporary processing during alignment
# 22a93f2 Merge branch 'main' into main
# 289eadf fix a merge error.
# ffbc736 change the docstrings and comments to English
# 36d2622 feat: add Latvian align model
# 4db8390 feat: add Tagalog (tl - Filipino) Phoneme-based ASR Model
# 0d9807a feat: add Basque alignment model
# 8c58c54 Revert "feat: add Basque alignment model"
# e7712f4 refactor: update import statements to use explicit module paths

COMMITS=(
    "v3.3.2:BASELINE v3.3.2"
    "69281f3:support timestamps for numbers"
    "65b2332:make align a bit faster"
    "4ebfb07:make no beam consistent with backtrack"
    "f286e7f:refactor type hints and imports"
    "024bc84:consolidate segment data handling"
    "2f93e02:add SegmentData type"
    "289eadf:fix a merge error"
    "ffbc736:change docstrings to English"
    "36d2622:add Latvian align model"
    "e7712f4:update import paths"
)

for entry in "${COMMITS[@]}"; do
    IFS=':' read -r ref label <<< "$entry"
    test_commit "$ref" "$label"
done

echo ""
echo "=== Restoring v3.7.6 alignment.py ==="
cp "$WHISPERX_DIR/alignment_v376.py" "$WHISPERX_DIR/alignment.py"
rm -rf "$WHISPERX_DIR/__pycache__"
echo "Done."
```
</details>

**Comment by PavelPanjaya** [comment]
@Barabazs 

> Can you confirm whether you used this exact workflow and audio on v3.3.2 and it produced a correct alignment?

Yes, I used the same code I shared to produce timestamps with all versions starting from `3.3.2` up to `3.8.1`.
With `3.3.2` I got correct timestamps on the same input.

**Comment by Barabazs** [comment]
Thanks @PavelPanjaya I managed to reproduce it and properly revert the issues. You can test it on https://github.com/m-bain/whisperX/tree/fix/alignment-1220 if you'd like to.

**Comment by Barabazs** [comment]
Thank you all for the helpful comments, audio samples and scripts!

**Comment by wrichert** [comment]
Awesome, thanks so much @Barabazs !! What are the next steps now? Will you push another release on the main repo so that we can update our pip version reqs?

**Comment by Barabazs** [comment]
@wrichert latest release + backports have been published to pypi
